src/persist-client/src/fetch.rs

```diff
@@ -493,14 +493,35 @@ where
     type Item = ((Result<K, String>, Result<V, String>), T, D);
 
     fn next(&mut self) -> Option<Self::Item> {
-        while let Some((k, v, mut t, d)) = self.part.next() {
+        while let Some((k, v, mut t, d)) = self.part_cursor.pop(&self.part) {
             if !self.ts_filter.filter_ts(&mut t) {
                 continue;
             }
 
+            let mut d = D::decode(d);
+
+            // If `filter_ts` advances our timestamp, we may end up with the same K, V, T in successive
+            // records. If so, opportunistically consolidate those out.
+            while let Some((k_next, v_next, mut t_next, d_next)) = self.part_cursor.peek(&self.part)
+            {
+                if (k, v) != (k_next, v_next) {
+                    break;
+                }
+
+                if !self.ts_filter.filter_ts(&mut t_next) {
+                    break;
+                }
+                if t != t_next {
+                    break;
+                }
+
+                // All equal... consolidate!
+                self.part_cursor.idx += 1;
+                d.plus_equals(&D::decode(d_next));
+            }
+
```

Comment:
can we do an `if d.is_zero() { continue; }` (and include coverage in the test)?

Reply:
Yes!